File: swarm_control/p5js/cluster_detector.js
// クラスタリングと輪郭検出（デッドロック解放）

class ClusterDetector {
    constructor() {
        this.clusters = [];
        this.contours = []; // 各クラスタの輪郭（凸包）
    }
    
    // 距離ベースクラスタリング（100mm以内のロボットをグループ化）
    detectClusters(robots) {
        this.clusters = [];
        const processed = new Set();
        const clusterDistance = Params.clustering.clusterDistance;
        
        for (let i = 0; i < robots.length; i++) {
            if (processed.has(i)) continue;
            
            // 新しいクラスタを開始
            const cluster = {
                indices: [i],
                positions: [{ x: robots[i].x, y: robots[i].y }],
                centerX: robots[i].x,
                centerY: robots[i].y
            };
            
            processed.add(i);
            
            // 近接ロボットを収集
            let changed = true;
            while (changed) {
                changed = false;
                for (let j = 0; j < robots.length; j++) {
                    if (processed.has(j)) continue;
                    
                    // クラスタ内のいずれかのロボットと近接しているかチェック
                    for (const idx of cluster.indices) {
                        const dx = robots[j].x - robots[idx].x;
                        const dy = robots[j].y - robots[idx].y;
                        const dist = Math.hypot(dx, dy);
                        
                        if (dist < clusterDistance) {
                            cluster.indices.push(j);
                            cluster.positions.push({ x: robots[j].x, y: robots[j].y });
                            processed.add(j);
                            changed = true;
                            break;
                        }
                    }
                }
            }
            
            // クラスタの中心を計算
            if (cluster.positions.length > 0) {
                let sumX = 0, sumY = 0;
                for (const pos of cluster.positions) {
                    sumX += pos.x;
                    sumY += pos.y;
                }
                cluster.centerX = sumX / cluster.positions.length;
                cluster.centerY = sumY / cluster.positions.length;
            }
            
            // クラスタが2つ以上のロボットを含む場合のみ追加（デッドロックの可能性）
            if (cluster.indices.length >= 2) {
                this.clusters.push(cluster);
            }
        }
        
        return this.clusters;
    }
    
    // 凸包（convex hull）を計算（Graham scan風の簡易版）
    computeConvexHull(points) {
        if (points.length < 3) {
            // 点が3つ未満の場合はそのまま返す
            return points;
        }
        
        // 最も下（yが最大、同じならxが最小）の点を見つける
        let bottomIdx = 0;
        for (let i = 1; i < points.length; i++) {
            if (points[i].y > points[bottomIdx].y ||
                (points[i].y === points[bottomIdx].y && points[i].x < points[bottomIdx].x)) {
                bottomIdx = i;
            }
        }
        
        // 最も下の点を最初に移動
        const bottomPoint = points[bottomIdx];
        const sortedPoints = [bottomPoint];
        
        // 残りの点を角度でソート（簡易版：距離と角度で判定）
        const otherPoints = points.filter((_, idx) => idx !== bottomIdx);
        otherPoints.sort((a, b) => {
            const angleA = Math.atan2(a.y - bottomPoint.y, a.x - bottomPoint.x);
            const angleB = Math.atan2(b.y - bottomPoint.y, b.x - bottomPoint.x);
            if (Math.abs(angleA - angleB) < 0.001) {
                // 角度が同じ場合は距離で比較
                const distA = Math.hypot(a.x - bottomPoint.x, a.y - bottomPoint.y);
                const distB = Math.hypot(b.x - bottomPoint.x, b.y - bottomPoint.y);
                return distA - distB;
            }
            return angleA - angleB;
        });
        
        sortedPoints.push(...otherPoints);
        
        // Graham scan風のアルゴリズム（簡易版）
        const hull = [sortedPoints[0], sortedPoints[1]];
        
        for (let i = 2; i < sortedPoints.length; i++) {
            const point = sortedPoints[i];
            
            // 右回り（時計回り）の場合は前の点を削除
            while (hull.length >= 2) {
                const p1 = hull[hull.length - 2];
                const p2 = hull[hull.length - 1];
                
                // 外積で右回りか判定
                const cross = (p2.x - p1.x) * (point.y - p1.y) - (p2.y - p1.y) * (point.x - p1.x);
                if (cross <= 0) {
                    hull.pop();
                } else {
                    break;
                }
            }
            
            hull.push(point);
        }
        
        return hull;
    }
    
    // 輪郭（contour）を推定（各クラスタの凸包を計算）
    estimateContours(robots) {
        this.contours = [];
        
        // クラスタを検出
        const clusters = this.detectClusters(robots);
        
        // 各クラスタの凸包を計算
        for (const cluster of clusters) {
            const hull = this.computeConvexHull(cluster.positions);
            this.contours.push({
                cluster: cluster,
                hull: hull,
                centerX: cluster.centerX,
                centerY: cluster.centerY
            });
        }
        
        return this.contours;
    }
    
    // ロボットがクラスタ内にいるかチェック
    isInCluster(robot, cluster) {
        for (const idx of cluster.indices) {
            if (robot.index === idx) {
                return true;
            }
        }
        return false;
    }
    
    // ロボットから輪郭の外側方向を計算
    getEscapeDirection(robot, contour) {
        // ロボットがクラスタ内にいるかチェック
        if (!this.isInCluster(robot, contour.cluster)) {
            return null;
        }
        
        // 輪郭（凸包）の各点から、ロボットへの方向を計算
        // 最も外側方向（輪郭から離れる方向）を選択
        let bestDirection = null;
        let maxDistance = -Infinity;
        
        for (let i = 0; i < contour.hull.length; i++) {
            const p1 = contour.hull[i];
            const p2 = contour.hull[(i + 1) % contour.hull.length];
            
            // 輪郭の辺の中点
            const midX = (p1.x + p2.x) / 2;
            const midY = (p1.y + p2.y) / 2;
            
            // ロボットから輪郭の中点への方向
            const dx = midX - robot.x;
            const dy = midY - robot.y;
            const dist = Math.hypot(dx, dy);
            
            if (dist > maxDistance) {
                maxDistance = dist;
                // 外側方向（輪郭から離れる方向）を計算
                // 輪郭の法線方向を計算（簡易版：ロボットから輪郭への方向の逆）
                const angle = Math.atan2(dy, dx);
                const normalAngle = angle + Math.PI; // 180度回転（外側方向）
                
                bestDirection = {
                    x: Math.cos(normalAngle),
                    y: Math.sin(normalAngle),
                    strength: Params.clustering.escapeForce * (1.0 + dist / 100.0) // 距離に応じて強度を調整
                };
            }
        }
        
        return bestDirection;
    }
}

File: toio_control/pc_tools/swarm_control/src/algorithm/flocking/cluster_detector.cpp
#include "cluster_detector.hpp"
#include "../agent/robot_agent.hpp"
#include "../../utils/utils.hpp"
#include <algorithm>
#include <cmath>
#include <set>

namespace swarm_control {

ClusterDetector::ClusterDetector() = default;

std::vector<Cluster> ClusterDetector::detectClusters(
    const std::vector<RobotAgent*>& robots) {
    
    clusters_.clear();
    std::set<int> processed;
    const double clusterDistance = params::g_params.clustering.clusterDistance;
    
    for (size_t i = 0; i < robots.size(); ++i) {
        if (processed.count(static_cast<int>(i)) > 0) continue;
        
        // 新しいクラスタを開始
        Cluster cluster;
        cluster.indices.push_back(static_cast<int>(i));
        auto pos = robots[i]->getPosition();
        cluster.positions.push_back({pos.x, pos.y});
        cluster.centerX = pos.x;
        cluster.centerY = pos.y;
        
        processed.insert(static_cast<int>(i));
        
        // 近接ロボットを収集
        bool changed = true;
        while (changed) {
            changed = false;
            for (size_t j = 0; j < robots.size(); ++j) {
                if (processed.count(static_cast<int>(j)) > 0) continue;
                
                // クラスタ内のいずれかのロボットと近接しているかチェック
                for (int idx : cluster.indices) {
                    auto pos1 = robots[idx]->getPosition();
                    auto pos2 = robots[j]->getPosition();
                    const double dist = utils::distance(pos1.x, pos1.y, pos2.x, pos2.y);
                    
                    if (dist < clusterDistance) {
                        cluster.indices.push_back(static_cast<int>(j));
                        cluster.positions.push_back({pos2.x, pos2.y});
                        processed.insert(static_cast<int>(j));
                        changed = true;
                        break;
                    }
                }
            }
        }
        
        // クラスタの中心を計算
        if (!cluster.positions.empty()) {
            double sumX = 0.0, sumY = 0.0;
            for (const auto& pos : cluster.positions) {
                sumX += pos.x;
                sumY += pos.y;
            }
            cluster.centerX = sumX / cluster.positions.size();
            cluster.centerY = sumY / cluster.positions.size();
        }
        
        // クラスタが2つ以上のロボットを含む場合のみ追加
        if (cluster.indices.size() >= 2) {
            clusters_.push_back(cluster);
        }
    }
    
    return clusters_;
}

double ClusterDetector::crossProduct(const Position& p1, const Position& p2, 
                                     const Position& p3) const {
    return (p2.x - p1.x) * (p3.y - p1.y) - (p2.y - p1.y) * (p3.x - p1.x);
}

std::vector<Position> ClusterDetector::computeConvexHull(
    const std::vector<Position>& points) const {
    
    if (points.size() < 3) {
        return points;
    }
    
    // 最も下（yが最大、同じならxが最小）の点を見つける
    size_t bottomIdx = 0;
    for (size_t i = 1; i < points.size(); ++i) {
        if (points[i].y > points[bottomIdx].y ||
            (std::abs(points[i].y - points[bottomIdx].y) < 0.001 && 
             points[i].x < points[bottomIdx].x)) {
            bottomIdx = i;
        }
    }
    
    // 最も下の点を最初に移動
    const Position bottomPoint = points[bottomIdx];
    std::vector<Position> sortedPoints = {bottomPoint};
    
    // 残りの点を角度でソート
    std::vector<Position> otherPoints;
    for (size_t i = 0; i < points.size(); ++i) {
        if (i != bottomIdx) {
            otherPoints.push_back(points[i]);
        }
    }
    
    std::sort(otherPoints.begin(), otherPoints.end(),
        [&bottomPoint](const Position& a, const Position& b) {
            const double angleA = std::atan2(a.y - bottomPoint.y, a.x - bottomPoint.x);
            const double angleB = std::atan2(b.y - bottomPoint.y, b.x - bottomPoint.x);
            if (std::abs(angleA - angleB) < 0.001) {
                const double distA = utils::distance(a.x, a.y, bottomPoint.x, bottomPoint.y);
                const double distB = utils::distance(b.x, b.y, bottomPoint.x, bottomPoint.y);
                return distA < distB;
            }
            return angleA < angleB;
        });
    
    sortedPoints.insert(sortedPoints.end(), otherPoints.begin(), otherPoints.end());
    
    // Graham scan風のアルゴリズム
    std::vector<Position> hull = {sortedPoints[0], sortedPoints[1]};
    
    for (size_t i = 2; i < sortedPoints.size(); ++i) {
        const Position& point = sortedPoints[i];
        
        // 右回り（時計回り）の場合は前の点を削除
        while (hull.size() >= 2) {
            const Position& p1 = hull[hull.size() - 2];
            const Position& p2 = hull[hull.size() - 1];
            
            const double cross = crossProduct(p1, p2, point);
            if (cross <= 0.0) {
                hull.pop_back();
            } else {
                break;
            }
        }
        
        hull.push_back(point);
    }
    
    return hull;
}

std::vector<Contour> ClusterDetector::estimateContours(
    const std::vector<RobotAgent*>& robots) {
    
    contours_.clear();
    
    // クラスタを検出
    const auto clusters = detectClusters(robots);
    
    // 各クラスタの凸包を計算
    for (const auto& cluster : clusters) {
        const auto hull = computeConvexHull(cluster.positions);
        Contour contour;
        contour.cluster = cluster;
        contour.hull = hull;
        contour.centerX = cluster.centerX;
        contour.centerY = cluster.centerY;
        contours_.push_back(contour);
    }
    
    return contours_;
}

bool ClusterDetector::isInCluster(const RobotAgent* robot, const Cluster& cluster) const {
    const int robotIndex = robot->getIndex();
    return std::find(cluster.indices.begin(), cluster.indices.end(), robotIndex) 
           != cluster.indices.end();
}

std::optional<EscapeDirection> ClusterDetector::getEscapeDirection(
    const RobotAgent* robot, const Contour& contour) const {
    
    // ロボットがクラスタ内にいるかチェック
    if (!isInCluster(robot, contour.cluster)) {
        return std::nullopt;
    }
    
    auto robotPos = robot->getPosition();
    
    // 輪郭（凸包）の各辺から、ロボットへの方向を計算
    EscapeDirection bestDirection;
    double maxDistance = -1.0;
    
    for (size_t i = 0; i < contour.hull.size(); ++i) {
        const Position& p1 = contour.hull[i];
        const Position& p2 = contour.hull[(i + 1) % contour.hull.size()];
        
        // 輪郭の辺の中点
        const double midX = (p1.x + p2.x) / 2.0;
        const double midY = (p1.y + p2.y) / 2.0;
        
        // ロボットから輪郭の中点への方向
        const double dx = midX - robotPos.x;
        const double dy = midY - robotPos.y;
        const double dist = utils::distance(robotPos.x, robotPos.y, midX, midY);
        
        if (dist > maxDistance) {
            maxDistance = dist;
            // 外側方向（輪郭から離れる方向）を計算
            const double angle = std::atan2(dy, dx);
            const double normalAngle = angle + M_PI; // 180度回転
            
            bestDirection.x = std::cos(normalAngle);
            bestDirection.y = std::sin(normalAngle);
            bestDirection.strength = params::g_params.clustering.escapeForce * 
                                     (1.0 + dist / 100.0);
        }
    }
    
    return bestDirection;
}

}  // namespace swarm_control

